pkg/upgrade/upgrademanager/manager.go

```diff
@@ -354,125 +380,197 @@ func (m *Manager) Migrate(
-			// presence of new v21.1 or v21.2 Nodes being added to the cluster.
-			//
-			//   - All Nodes are running v21.1
-			//   - All Nodes are rolled into v21.2 binaries, but with active
-			//     cluster version still as v21.1
-			//   - The first version bump will be into v21.2-1(fence), see the
-			//     upgrade manager above for where that happens
-			//
-			// Then concurrently:
-			//
-			//   - A new node is added to the cluster, but running binary v21.1
-			//   - We try bumping the cluster gates to v21.2-1(fence)
-			//
-			// If the v21.1 Nodes manages to sneak in before the version bump,
-			// it's fine as the version bump is a no-op one (all fence versions
-			// are). Any subsequent bumps (including the "actual" one bumping to
-			// v21.2) will fail during the validation step where we'll first
-			// check to see that all Nodes are running v21.2 binaries.
-			//
-			// If the v21.1 node is only added after v21.2-1(fence) is active,
-			// it won't be able to actually join the cluster (it'll be prevented
-			// by the join RPC).
-			//
-			// All of which is to say that once we've seen the node list
-			// stabilize (as UntilClusterStable enforces), any new nodes that
-			// can join the cluster will run a release that support the fence
-			// version, and by design also supports the actual version (which is
-			// the direct successor of the fence).
-			fenceVersion := upgrade.FenceVersionFor(ctx, cv)
-			if err := bumpClusterVersion(ctx, m.deps.Cluster, fenceVersion); err != nil {
+		// In the case where we're upgrading secondary tenants there's an extra
+		// dance that must be performed. After we write the fence version to
+		// the settings table we must validate that the set of sql servers
+		// running, matches those that were present when we performed the bump
+		// above. This is to handle the case where a sql server with an old
+		// binary starts up in between the bump and the persistence of the fence
+		// version. In that case, the sql server will be permitted to startup,
+		// but the upgrade can not continue. The retry logic here will detect
+		// the sql server with the old binary, and the upgrade will fail. If
+		// instead, the newly started sql server has a binary version which is
+		// upgrade-compatible, the retry will detect that and the upgrade will
+		// proceed. Note that we only need to do this extra dance when we write
+		// the first fence of a given upgrade process, since for subsequent
+		// fences, the too-low-binary sql server will be prevented from starting
+		// by the check in preStart.
+		if !writtenFirstFence {
+			var err error
+			for err = updateSystemVersionSetting(ctx, cv, validate); err != nil && strings.Contains(err.Error(), "new SQL servers added during migration"); {
+				// Revalidate to ensure that added SQL server is at an allowable level
+				if err := validateTargetClusterVersion(ctx, m.deps.Cluster, fenceVersion); err != nil {
```

Comment:
what benefit do we get out of issuing the validate RPC again and then issuing the bump RPC given we validate inside bump? At this point the fence version has been written to the settings table, so there is no going backwards, right?